### PR TITLE
Enable compression in netty

### LINKED ISSUE #8486: Enable compression in netty
**Is your feature request related to a problem? Please describe.**

More and more usecases are coming up where Zeebe is deployed across multiple data centers. That means, the communication latency between two zeebe brokers can be high.

When there is high latency between nodes, we have previously observed
* [Frequent leader changes ](https://jira.camunda.com/browse/SUPPORT-9659)
* And/OR lower throughput

While we expect that higher latency will impact the commit latency and this overall process execution time and throughput, frequent leader change is not acceptable. One way to prevent leader change was to increase the election timeout. This means that failover time is also higher, as it take longer to start an election when the current leader dies.

One of the causes for the frequent leader change is that the bandwidth limitation of TCP when the RTT is higher.
https://accedian.com/blog/measuring-network-performance-latency-throughput-packet-loss/
image

Since we are sending a lot of data over network, the latency has a big impact on replication throughput. This cause a lot of requests to timeout and this causes a leader election.

**Describe the solution you'd like**

* Compress messages that are sent over network. It is easy to plugin a compression algorithm to netty. 
We can enable compression by adding the following lines when creating a channel (`BasicServer/ClientChannelInitializer#initChannel`
```
        channel.pipeline().addLast(ZlibCodecFactory.newZlibEncoder(ZlibWrapper.GZIP));
        channel.pipeline().addLast(ZlibCodecFactory.newZlibDecoder(ZlibWrapper.GZIP));
```
There are also other compression algorithms available.

**Describe alternatives you've considered**
* To prevent freq leader change, we can increase the election timeout. But this would mean failover time is high as it takes longer to detect a leader failure.
* Compress records before writing to disk. Raft should replicate already compressed records. This may have an additional benefit that the disk i/o is reduced.

**Additional context**
See comment for the benchmark result from a prototype.

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

* Add two compression algorithms to netty (gzip, snappy)
* Compression can be enabled by configuration (default = disabled)
* When compression is enabled all messages sent between all brokers and gateway will be compressed.

In the spike, we observed that compression increased the throughput when network latency is high. In no latency test, we found that enabling compression has not significant performance impact. 

## Related issues

closes #8486

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark 

Documentation: 
* [x] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by deepthidevaki** [comment]
Throughput Snappy
image

Throughput Gzip
image

Compression Snappy:
image

Compression GZip
image

**Comment by npepinpe** [comment]
The snappy one seems within normal variation imo, unless that was the max it ever reached...and even then, I would still assume it was fairly similar.

Did you see Ole's last comment about using zstd instead of gzip?

**Comment by deepthidevaki** [comment]
> 

Yes. I would suggest to add zstd as a follow up.

File: gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/CompressionCfg.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    return algorithm;
+  }
+
+  public void setAlgorithm(final CompressionAlgorithm algorithm) {
+    this.algorithm = algorithm;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(enableCompression, algorithm);
```
Comment (oleschoenburg) [inline gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/CompressionCfg.java:37]: :upside_down_face: strictly speaking, this isn't the IntelliJ default style: https://github.com/camunda-cloud/zeebe/wiki/Code-Style#equalshashcode-implementations

Same goes for `MembershipCfg`

**Review by lenaschoenburg — APPROVED** [review]
Looks great! :+1: 

:wrench: Personally, I wouldn't mind a different configuration scheme where we only offer `zeebe.broker.cluster.messageCompression = none|gzip|snappy` with none being the default. 

:thought_balloon: : Do we want an end-to-end test for this? It's similar to the new TLS feature where we also don't have an e2e test (I think) so it might be fine but at least to me it's not immediately obvious that `NettyMessagingServiceCompressionTest` ensures this will work (and keep working) for broker to broker and gateway communication.

**Comment by npepinpe** [comment]
(just an FYI, we do have a QA test for cluster TLS called `SecureClusteredMessagingIT` :wink:)

File: dist/src/main/config/broker.standalone.yaml.template [hunk 2]
```diff
@@ -325,6 +325,20 @@
         # This setting can also be overridden using the environment variable ZEEBE_BROKER_CLUSTER_MEMBERSHIP_SYNCINTERVAL
         # syncInterval: 10s
 
+      # messageCompression:
+        # When compression is enabled, all messages between the brokers and between the broker and
+        # the gateway is compressed using the configured compression algorithm. This feature is useful
+        # when the network latency between the brokers is very high (for example when the brokers are deployed in different data centers).
+        # When latency is high, the network bandwidth is severely reduced. Hence enabling compression helps to improve the throughput.
+        # Note: When there is no latency enabling this may have a performance impact.
+        # Note: When this flag is enables, you should also enable compression in standalone gateway configuration.
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply dist/src/main/config/broker.standalone.yaml.template:334]: Yes. It is a "must". I will update the doc.
I don't know exactly, but I remember seeing an exception from the decoder.

File: gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/CompressionCfg.java (hunk 1 above)
Comment (deepthidevaki) [reply gateway/src/main/java/io/camunda/zeebe/gateway/impl/configuration/CompressionCfg.java:37]: :+1: Other configuration classes uses this format though. I will update the new code to follow the code style.

**Comment by deepthidevaki** [comment]
> Looks great! +1
> 
> wrench Personally, I wouldn't mind a different configuration scheme where we only offer `zeebe.broker.cluster.messageCompression = none|gzip|snappy` with none being the default.

I like it. I will change it. 

> 
> thought_balloon : Do we want an end-to-end test for this? It's similar to the new TLS feature where we also don't have an e2e test (I think) so it might be fine but at least to me it's not immediately obvious that `NettyMessagingServiceCompressionTest` ensures this will work (and keep working) for broker to broker and gateway communication.

I can add a integration test that verifies "it works" when compression is enabled. Is that enough? I don't think there is a way to check if messages are actually compressed.

**Comment by npepinpe** [comment]
I have a dumb question, since we talked about this with Ole yesterday. The Netty channel pipeline modifies itself - once the protocol handshake is finished, it removes that handler, and adds other handlers at the end of the pipeline. Does that affect the compression? How does the pipeline handle it? Does it mean after the handshake, the new things are added at the end of the pipeline after the compression encoding? :thinking:

**Comment by deepthidevaki** [comment]
> I have a dumb question, since we talked about this with Ole yesterday. The Netty channel pipeline modifies itself - once the protocol handshake is finished, it removes that handler, and adds other handlers at the end of the pipeline. Does that affect the compression? How does the pipeline handle it? Does it mean after the handshake, the new things are added at the end of the pipeline after the compression encoding? thinking I guess not, but I don't understand how it works sweat_smile

I don't know what you are referring to. I only know that the handlers are executed in the specific order (one order for inbound and reverse order for outbound).

**Comment by deepthidevaki** [comment]
@oleschoenburg I changed the configuration option to `cluster.messageCompression = NONE | GZIP | SNAPPY`. Since there are lot of changes, I squashed the commits and force-pushed them. Could you please review it again?

**Comment by npepinpe** [comment]
If you look at the `ClientHandshakeHandlerAdapter` (which is originally added to the pipeline before the compression stuff), in `activateProtocolVersion` it will remove itself, and then add new encoder/decoder and message dispatcher handlers to the pipeline. I'm wondering how this interacts with the compression handlers.

So originally, the pipeline is:

Inbound: `TLS` -> `ClientHandshakeHandlerAdapter` -> `CompressionDecoder`
Outbound: `CompressionEncoder` -> `ClientHandshakeHandlerAdapter` -> `TLS`

Then the pipeline is modified after the initial handshake:

Inbound: `TLS` -> `CompressionDecoder` -> `ProtocolDecoder` -> `MessageDispatcher`
Outbound: `ProtocolEncoder` -> `CompressionEncoder` -> `TLS`

The final pipeline looks good and what I would expect, but the initial one looks weird - and I have a feeling we ended up like that unexpectedly? I always have a hard time wrapping my head around this self modifying pipeline. It might be worth having a follow up issue where the pipeline is fixed but stateful (a bit like the SSL handler, where it knows the handshake was performed so it can now do other stuff). Also, I guess for the initial outbound pipeline, the handshake adapter should be before the compression encoder? I'm not sure why that didn't break anything :man_shrugging:

**Review by lenaschoenburg — APPROVED** [review]
:tada: LGTM, thanks for updating the config scheme and writing the e2e test.

**Comment by oleschoenburg** [comment]
> The final pipeline looks good and what I would expect, but the initial one looks weird

Hm, I guess that just means that the initial handshake does not use compression, which makes sense. I suppose it'd be also OK to only install the compression handlers after a successful handshake?

**Comment by npepinpe** [comment]
It still seems like the end result is unexpected? :smile: I can create a follow up issue to sort it out and make it clearer.

**Comment by deepthidevaki** [comment]
@npepinpe Please create a follow up issue. I don't really understand the problem you found. 
I guess I can merge this, as it seems to work? :slightly_smiling_face:

**Comment by deepthidevaki** [comment]
bors merge

**Comment by npepinpe** [comment]
I think the fact none of us understand it is the real problem :smile:

**Pull request merged**